sympy/functions/elementary/exponential.py

```diff
@@ -1055,6 +1058,8 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
             cdir = arg0.dir(x, cdir)
         if x0.is_real and x0.is_negative and im(cdir).is_negative:
             return self.func(x0) - 2*I*S.Pi
+        if isinstance(arg, AccumBounds) and arg.min < S.Zero:
+            return S.NaN
```

Comment:
A better place for this might be `_eval`. If arg.min is 0, then `AccumBounds(-oo, arg.max)`.

Reply:
Cool , would give this a thought !

Reply:
I am a bit confused about the `arg` your using . Our expression is `log(cos(1/x))` . So the `arg` which `eval` uses is `self.args[0] `i.e `cos(1/x)` . But the `arg` we use in `_eval_as_leading_term` is `cos(1/x).as_leading_term(x)` which is `AccumBounds(-1, 1)`. So as per what you've written if we deal with this in eval we must be using something like `arg.as_leading_term(x).min == 0`

Reply:
Maybe we could end up with a block like this in `eval`
```
        leading_term = arg.as_leading_term(*list(arg.free_symbols))
        if isinstance(leading_term, AccumBounds):
            from sympy import oo
            if leading_term.min < S.Zero:
                return AccumBounds(-oo, leading_term.max)
```
If we include this we end up with something as follows
```
>>> log(cos(1/x))
AccumBounds(-oo, 1)
```

Reply:
Returning min = -oo only applies when the argument AccumBounds has min == 0. Is it is < 0, then nan should be returned. Currently, there is only the min > 0 case in `eval`. The two new cases should be added.

Reply:
Thanks for this suggestions , works better now . I also wanted your assistance on another thing which is the limit we are fixing but for the `tan/cot/sec/csc` cases . I realized that wolfram returns indefinite for 
`limit(sin(1/x)**x, x, 0)` and `limit(cos(1/x)**x, x, 0)` but for similar cases of `tan/cot/sec/csc` , it returns 1 . I was a bit confused but when after applying your suggestion it seems that's how it is supposed to be .
The diff I used 
```
diff --git a/sympy/functions/elementary/exponential.py b/sympy/functions/elementary/exponential.py
index 2030458b83..9095d9b5f5 100644
--- a/sympy/functions/elementary/exponential.py
+++ b/sympy/functions/elementary/exponential.py
@@ -703,6 +703,10 @@ def eval(cls, arg, base=None):
         elif isinstance(arg, AccumBounds):
             if arg.min.is_positive:
                 return AccumBounds(log(arg.min), log(arg.max))
+            elif arg.min.is_zero:
+                return AccumBounds(S.NegativeInfinity, log(arg.max))
+            elif arg.min.is_negative:
+                return S.NaN
             else:
                 return
         elif isinstance(arg, SetExpr):
@@ -1058,8 +1062,8 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
             cdir = arg0.dir(x, cdir)
         if x0.is_real and x0.is_negative and im(cdir).is_negative:
             return self.func(x0) - 2*I*S.Pi
-        if isinstance(arg, AccumBounds) and arg.min < S.Zero:
-            return S.NaN
+        if isinstance(arg, AccumBounds):
+            return log(arg)
         return self.func(arg)
```
Now I get 
```
>>> limit(sin(1/x)**x, x, 0) # returning self because exp(x*log(sin(1/x))).as_leading_term is nan
Limit(sin(1/x)**x, x, 0)
>>> limit(cos(1/x)**x, x, 0)
Limit(cos(1/x)**x, x, 0)
>>> limit(tan(1/x)**x, x, 0) # returning 1 because exp(x*log(tan(1/x))).as_leading_term(x) is 1
1
>>> limit(cot(1/x)**x, x, 0)
1
>>> limit(csc(1/x)**x, x, 0)
1
>>> limit(sec(1/x)**x, x, 0)
1
```
This is because 
```
>>> log(AccumBounds(-1, 1))
nan
>>> log(AccumBounds(-1, oo))
nan
>>> log(AccumBounds(-oo, oo))
log(AccumBounds(-oo, oo))
```

Wolfram pics for reference
![image](https://user-images.githubusercontent.com/87052487/154538432-b347d4de-8f71-4212-a21c-f55e2686bbae.png)
![image](https://user-images.githubusercontent.com/87052487/154538528-c205c0f5-d7c7-4081-ab68-6a869aa19c94.png)
![image](https://user-images.githubusercontent.com/87052487/154538628-9bd3e9a4-0aa1-4b2a-9445-8521670ed43f.png)

Reply:
Hence I wanted to know that now that our changes reported above correctly work for `tan/cot/sec/csc` and return `1 ` , I hope there isn't anything suspicious or doubtful here ?? I hope the results shown above are the ones we need to return .

Reply:
If this is what we expect, I'll make the commit soon .